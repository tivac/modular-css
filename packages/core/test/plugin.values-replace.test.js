"use strict";

var path   = require("path"),
    
    postcss = require("postcss"),
    dedent  = require("dedent"),

    plugin     = require("../plugins/values-replace.js"),
    local      = require("../plugins/values-local.js"),
    composed   = require("../plugins/values-composed.js"),
    exported   = require("../plugins/values-export.js"),
    namespaced = require("../plugins/values-namespaced.js"),
    
    resolve = require("../lib/resolve.js").resolve;

describe("/plugins", function() {
    describe("/values-replace.js", function() {
        it("should noop without values to replace", function() {
            expect(
                postcss([ plugin ]).process(".wooga { color: red; }").css
            )
            .toMatchSnapshot();
        });

        describe("local values", function() {
            var processor = postcss([ local, exported, plugin ]);

            function process(css) {
                return processor.process(css, {
                    resolve,
                    
                    from  : "file",
                    files : {
                        file : {
                            values : false
                        }
                    }
                });
            }
            
            it("should replace values in declarations", function() {
                expect(
                    process(dedent(`
                        @value color: red;
                        
                        .wooga {
                            color: color;
                        }
                    `)).css
                )
                .toMatchSnapshot();
            });

            it("should replace value references in @value declarations", function() {
                expect(
                    process(dedent(`
                        @value color: red;
                        @value value: color;
                        
                        .wooga {
                            color: value;
                        }
                    `)).css
                )
                .toMatchSnapshot();

                expect(
                    process(dedent(`
                        @value red: #F00;
                        @value color: red;
                        @value value: color;
                        
                        .wooga {
                            color: value;
                        }
                    `)).css
                )
                .toMatchSnapshot();
            });

            it("should replace values in media queries", function() {
                expect(
                    process(dedent(`
                        @value small: (max-width: 599px);
                        
                        @media small { }
                    `)).css
                )
                .toMatchSnapshot();
            });

            it("should replace values surrounded by non-word characters (#245)", function() {
                expect(
                    process(dedent(`
                        @value one: 10px;
                        @value two: calc(one - 2px);
                        
                        .a {
                            height: two;
                            width: -one;
                            color: twoodle;
                        }
                    `)).css
                )
                .toMatchSnapshot();
            });
        });

        describe("composed values", function() {
            var css = postcss([ composed, exported, plugin ]);
            
            it("should replace values in declarations", function() {
                expect(
                    css.process(dedent(`
                        @value color from "./local.css";
                        .wooga {
                            color: color;
                        }
                    `), {
                        resolve,
                        
                        map   : false,
                        from  : path.resolve("./packages/core/test/specimens/in.css"),
                        files : {
                            // Composition source
                            [ path.resolve("./packages/core/test/specimens/in.css") ] : {},

                            // Composition target
                            [ path.resolve("./packages/core/test/specimens/local.css") ] : {
                                values : {
                                    color : {
                                        value  : "#F00",
                                        source : {}
                                    }
                                }
                            }
                        }
                    }).css
                )
                .toMatchSnapshot();
            });
        });

        describe("namespaced values", function() {
            var css = postcss([ namespaced, exported, plugin ]);
            
            it("should replace values in declarations", function() {
                expect(
                    css.process(dedent(`
                        @value * as colors from "./local.css";
                        .wooga {
                            color: colors.red;
                        }
                    `), {
                        resolve,

                        map   : false,
                        from  : path.resolve("./packages/core/test/specimens/in.css"),
                        files : {
                            // Composition source
                            [ path.resolve("./packages/core/test/specimens/in.css") ] : {},

                            // Composition target
                            [ path.resolve("./packages/core/test/specimens/local.css") ] : {
                                values : {
                                    red : {
                                        value  : "#F00",
                                        source : {}
                                    }
                                }
                            }
                        }
                    }).css
                )
                .toMatchSnapshot();
            });
        });

        describe("combined", function() {
            var css = postcss([ local, namespaced, composed, exported, plugin ]);

            it("should replace values in declarations", function() {
                expect(
                    css.process(dedent(`
                        @value * as colors from "./local.css";
                        @value red from "./local.css";
                        @value r: colors.red;
                        .wooga {
                            color: colors.red;
                            background: red;
                            border: 1px solid r;
                        }
                    `), {
                        resolve,
                        
                        map   : false,
                        from  : path.resolve("./packages/core/test/specimens/in.css"),
                        files : {
                            // Composition source
                            [ path.resolve("./packages/core/test/specimens/in.css") ] : {},

                            // Composition target
                            [ path.resolve("./packages/core/test/specimens/local.css") ] : {
                                values : {
                                    red : {
                                        value  : "#F00",
                                        source : {}
                                    }
                                }
                            }
                        }
                    }).css
                )
                .toMatchSnapshot();
            });
        });
    });
});
