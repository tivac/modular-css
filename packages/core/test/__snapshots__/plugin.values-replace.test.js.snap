// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/plugins /values-replace.js combined should replace values in declarations 1`] = `
".wooga {
    color: #F00;
    background: #F00;
    border: 1px solid #F00;
}"
`;

exports[`/plugins /values-replace.js composed values should replace values in declarations 1`] = `
".wooga {
    color: #F00;
}"
`;

exports[`/plugins /values-replace.js local values should replace value references in @value declarations 1`] = `
".wooga {
    color: red;
}"
`;

exports[`/plugins /values-replace.js local values should replace value references in @value declarations 2`] = `
".wooga {
    color: #F00;
}"
`;

exports[`/plugins /values-replace.js local values should replace values in declarations 1`] = `
".wooga {
    color: red;
}"
`;

exports[`/plugins /values-replace.js local values should replace values in media queries 1`] = `"@media (max-width: 599px) { }"`;

exports[`/plugins /values-replace.js local values should replace values surrounded by non-word characters (#245) 1`] = `
".a {
    height: calc(10px - 2px);
    width: -10px;
    color: twoodle;
}"
`;

exports[`/plugins /values-replace.js namespaced values should replace values in declarations 1`] = `
".wooga {
    color: #F00;
}"
`;

exports[`/plugins /values-replace.js should noop without values to replace 1`] = `".wooga { color: red; }"`;
